### ISSUE TITLE
Conflicting dependencies?

### ISSUE THREAD

**Issue by bzg** [plexus/chestnut#203]
Hi Arne,

I cannot add `[com.datomic/datomic-free "0.9.5561.50"]` to a chesnut project created with:

> ~$ lein new chestnut app +http-kit +no-poll +vanilla 

Simply adding the datomic-free library to the `project.clj` then running `lein repl` results in:

> Caused by: java.lang.IllegalArgumentException: No matching ctor found for class clj_logging_config.log4j.proxy$org.apache.log4j.WriterAppender$ff19274a

I added a few exclusions as suggested by `lein deps :tree` but this does not fix the main issue, which seems to stem from what logging library datomic-free (and pro btw) depends on.

Here is the complete output of `lein deps :tree` **with** the datomic-free library and the error.

```
 [bk/ring-gzip "0.2.1"]
 [clojure-complete "0.2.4" :exclusions [[org.clojure/clojure]]]
 [com.cemerick/piggieback "0.2.1" :scope "test"]
 [com.cognitect/transit-clj "0.8.300"]
   [com.cognitect/transit-java "0.8.324"]
     [com.fasterxml.jackson.core/jackson-core "2.8.7"]
     [org.msgpack/msgpack "0.6.12"]
       [com.googlecode.json-simple/json-simple "1.1.1" :exclusions [[junit]]]
       [org.javassist/javassist "3.18.1-GA"]
 [com.datomic/datomic-free "0.9.5561.50"]
   [com.datomic/datomic-lucene-core "3.3.0"]
   [com.google.guava/guava "18.0"]
   [com.h2database/h2 "1.3.171"]
   [commons-codec "1.10"]
   [org.apache.activemq/artemis-core-client "1.4.0" :exclusions [[org.jgroups/jgroups] [commons-logging]]]
     [io.netty/netty-all "4.0.39.Final"]
     [org.apache.activemq/artemis-commons "1.4.0"]
       [commons-beanutils "1.9.2"]
         [commons-collections "3.2.1"]
       [org.jboss.logging/jboss-logging "3.3.0.Final"]
     [org.apache.geronimo.specs/geronimo-json_1.0_spec "1.0-alpha-1"]
     [org.apache.johnzon/johnzon-core "0.9.4"]
   [org.apache.tomcat/tomcat-jdbc "7.0.27" :exclusions [[commons-logging]]]
     [org.apache.tomcat/tomcat-juli "7.0.27"]
   [org.clojure/tools.cli "0.3.5" :exclusions [[org.clojure/clojure]]]
   [org.codehaus.janino/commons-compiler-jdk "2.6.1"]
     [org.codehaus.janino/commons-compiler "2.6.1"]
   [org.fressian/fressian "0.6.5"]
   [org.slf4j/jcl-over-slf4j "1.7.7"]
   [org.slf4j/jul-to-slf4j "1.7.7"]
     [org.slf4j/slf4j-api "1.7.7"]
   [org.slf4j/log4j-over-slf4j "1.7.7" :scope "runtime"]
   [org.slf4j/slf4j-nop "1.7.7"]
 [com.stuartsierra/component "0.3.2"]
   [com.stuartsierra/dependency "0.2.0"]
 [compojure "1.6.0"]
   [clout "2.1.2"]
     [instaparse "1.4.0" :exclusions [[org.clojure/clojure]]]
   [medley "1.0.0"]
   [org.clojure/tools.macro "0.1.5"]
   [ring/ring-codec "1.0.1"]
 [environ "1.1.0"]
 [figwheel-sidecar "0.5.10" :scope "test" :exclusions [[com.google.guava/guava]]]
   [clj-stacktrace "0.2.8"]
   [hawk "0.2.11" :scope "test" :exclusions [[org.clojure/clojure]]]
     [net.incongru.watchservice/barbary-watchservice "1.0" :scope "test"]
       [net.java.dev.jna/jna "3.2.2" :scope "test"]
   [ring-cors "0.1.8" :scope "test" :exclusions [[ring/ring-core] [org.clojure/clojure]]]
   [simple-lein-profile-merge "0.1.4" :scope "test"]
   [strictly-specking-standalone "0.1.1" :scope "test"]
     [net.cgrand/parsley "0.9.3" :scope "test" :exclusions [[org.clojure/clojure]]]
       [net.cgrand/regex "1.1.0" :scope "test"]
     [net.cgrand/sjacket "0.1.1" :scope "test" :exclusions [[org.clojure/clojure] [net.cgrand/parsley]]]
 [figwheel "0.5.10" :scope "test" :exclusions [[com.google.guava/guava]]]
   [org.clojure/core.async "0.3.442" :scope "test" :exclusions [[org.clojure/tools.reader]]]
     [org.clojure/tools.analyzer.jvm "0.7.0" :scope "test"]
       [org.clojure/core.memoize "0.5.9" :scope "test"]
         [org.clojure/core.cache "0.6.5" :scope "test"]
           [org.clojure/data.priority-map "0.0.7" :scope "test"]
       [org.clojure/tools.analyzer "0.6.9" :scope "test"]
       [org.ow2.asm/asm-all "4.2" :scope "test"]
 [http-kit "2.2.0"]
 [lambdaisland/ring.middleware.logger "0.5.1"]
   [onelog "0.5.0"]
     [clj-logging-config "1.9.10" :exclusions [[log4j]]]
     [org.clojure/tools.logging "0.2.6"]
     [org.slf4j/slf4j-log4j12 "1.7.7"]
       [log4j "1.2.17"]
   [org.clojars.pjlegato/clansi "1.3.0"]
   [org.tobereplaced/mapply "1.0.0"]
 [lein-doo "0.1.7" :scope "test"]
   [doo "0.1.7" :scope "test"]
     [karma-reporter "0.1.0" :scope "test"]
 [org.clojure/clojure "1.8.0"]
 [org.clojure/clojurescript "1.9.562" :scope "provided" :exclusions [[com.google.guava/guava]]]
   [com.google.javascript/closure-compiler-unshaded "v20170521" :scope "provided"]
     [args4j "2.33" :scope "provided"]
     [com.google.code.findbugs/jsr305 "3.0.1" :scope "provided"]
     [com.google.code.gson/gson "2.7" :scope "provided"]
     [com.google.errorprone/error_prone_annotations "2.0.18" :scope "provided"]
     [com.google.javascript/closure-compiler-externs "v20170521" :scope "provided"]
     [com.google.jsinterop/jsinterop-annotations "1.0.0" :scope "provided"]
     [com.google.protobuf/protobuf-java "3.0.2" :scope "provided"]
   [org.clojure/data.json "0.2.6" :scope "provided"]
   [org.clojure/google-closure-library "0.0-20170519-fa0499ef" :scope "provided"]
     [org.clojure/google-closure-library-third-party "0.0-20170519-fa0499ef" :scope "provided"]
   [org.clojure/tools.reader "1.0.0-beta3" :scope "provided"]
   [org.mozilla/rhino "1.7R5" :scope "provided"]
 [org.clojure/tools.namespace "0.2.11"]
 [org.clojure/tools.nrepl "0.2.13" :scope "test"]
 [org.danielsz/system "0.4.0" :exclusions [[org.clojure/tools.namespace]]]
   [io.aviso/pretty "0.1.26"]
   [org.danielsz/lang-utils "0.1.0-20170423.140704-11"]
   [prismatic/schema "1.1.3"]
 [reloaded.repl "0.2.3" :scope "test"]
   [suspendable "0.1.1" :scope "test"]
 [ring/ring-defaults "0.3.0"]
   [javax.servlet/javax.servlet-api "3.1.0"]
   [ring/ring-anti-forgery "1.1.0"]
   [ring/ring-headers "0.3.0"]
   [ring/ring-ssl "0.3.0"]
 [ring "1.6.1"]
   [ring/ring-core "1.6.1"]
     [clj-time "0.11.0"]
       [joda-time "2.8.2"]
     [commons-fileupload "1.3.2"]
     [commons-io "2.5"]
     [crypto-equality "1.0.0"]
     [crypto-random "1.2.0"]
   [ring/ring-devel "1.6.1"]
     [hiccup "1.0.5"]
     [ns-tracker "0.3.1"]
       [org.clojure/java.classpath "0.2.3"]
   [ring/ring-jetty-adapter "1.6.1"]
     [org.eclipse.jetty/jetty-server "9.2.21.v20170120"]
       [org.eclipse.jetty/jetty-http "9.2.21.v20170120"]
         [org.eclipse.jetty/jetty-util "9.2.21.v20170120"]
       [org.eclipse.jetty/jetty-io "9.2.21.v20170120"]
   [ring/ring-servlet "1.6.1"]
```

Thanks in advance for your feedback!

**Comment by plexus** [comment]
Try adding the most recent version of clj-logging-config to your dependencies, which is `[clj-logging-config "1.9.12"]`. Alternatively pick whatever version datomic depends on. If it causes problems in the other direction you can always rip out `ring.middleware.logger`.

You can also try putting the exclusion on the other dependency, I find that leiningen's suggestions often lead to pulling in the older of the two, instead of the newer one. 

In case you haven't see it, this might help https://lambdaisland.com/episodes/resolving-clojure-dependency-conflicts

**Comment by bzg** [comment]
Thanks for the quick feedback! And for the link to the video, I had already seen it, very useful.

I tried adding the most recent version of `clj-logging-config` without success.  And several other ideas: in every scenario, I have a clean output for `lein deps :tree` but I stumble upon an error.

Finally, I tried to rip out `ring.middleware.logger` (and any reference to it in the code) but I get this error:

```
~$ lein clean
~$ lein repl
Exception in thread "main" java.lang.ExceptionInInitializerError
        at clojure.main.<clinit>(main.java:20)
Caused by: java.lang.NoSuchMethodError: com.google.common.base.Preconditions.checkState(ZLjava/lang/String;Ljava/lang/Object;)V, compiling:(closure.clj:101:1)
        at clojure.lang.Compiler$DefExpr.eval(Compiler.java:464)
        at clojure.lang.Compiler.eval(Compiler.java:6932)
        at clojure.lang.Compiler.load(Compiler.java:7379)
        at clojure.lang.RT.loadResourceScript(RT.java:372)
        at clojure.lang.RT.loadResourceScript(RT.java:363)
        at clojure.lang.RT.load(RT.java:453)
        at clojure.lang.RT.load(RT.java:419)
```

I'd very much like to use chestnut with datomic, even if I need to sacrifice logger for now...

**Comment by plexus** [comment]
Here's an example of Chestnut with datomic-free added, and the logger removed. Works fine over here

https://github.com/plexus/chestnut-sans-logger

**Comment by bzg** [comment]
It works fine here too, thanks a lot for taking the time to do this!

I did not try adding `[com.google.guava/guava "20.0"]` explicitely in `project.clj`, but I excluded this from datomic (which requires guava 18.0) and from figwheel (which requires guava 19.0), thinking the project would then use the clojurescript requirement (guava 20.0).

Is there any reason why adding `[com.google.guava/guava "20.0"]` does not yield the same dependencies than just preventing earlier versions from being pulled?
